File: native/bindings/endstone/inventory/player_inventory.cpp
#include "player_inventory.h"

#include "inventory.h"
#include "item_stack.h"

#include <endstone/inventory/player_inventory.h>
#include <endstone/player.h>

#include <optional>
#include <string>

namespace aegilex::native::inventory {

class PlayerInventory::impl {
  public:
    explicit impl(endstone::PlayerInventory *inventory) noexcept : inventory(inventory)
    {
    }

    endstone::PlayerInventory *inventory;
};

PlayerInventory::PlayerInventory(endstone::PlayerInventory *inventory) noexcept
    : impl(std::make_shared<class PlayerInventory::impl>(inventory))
{
}

endstone::PlayerInventory *PlayerInventory::native() const noexcept
{
    return impl == nullptr ? nullptr : impl->inventory;
}

std::unique_ptr<Inventory> PlayerInventory::asInventory() const
{
    try {
        return native() == nullptr ? std::unique_ptr<Inventory>()
                                   : std::make_unique<Inventory>(static_cast<endstone::Inventory *>(native()));
    }
    catch (...) {
        return {};
    }
}

std::int32_t PlayerInventory::getHeldItemSlot() const
{
    try {
        return impl->inventory->getHeldItemSlot();
    }
    catch (...) {
        return 0;
    }
}

void PlayerInventory::setHeldItemSlot(const std::int32_t slot) const
{
    try {
        if (slot >= 0 && slot <= 8) {
            impl->inventory->setHeldItemSlot(slot);
        }
    }
    catch (...) {
    }
}

namespace {

std::unique_ptr<ItemStack> get_stack(const std::optional<endstone::ItemStack> &item) noexcept
{
    try {
        return item.has_value() ? std::unique_ptr<ItemStack>(new ItemStack(*item)) : std::unique_ptr<ItemStack>();
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void clear_stack(endstone::PlayerInventory &inventory,
                 void (endstone::PlayerInventory::*setter)(std::optional<endstone::ItemStack>)) noexcept
{
    try {
        (inventory.*setter)(std::nullopt);
    }
    catch (...) {
    }
}

} // namespace

std::unique_ptr<ItemStack> PlayerInventory::getHelmet() const
{
    try {
        return get_stack(impl->inventory->getHelmet());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setHelmet(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setHelmet(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearHelmet() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setHelmet);
}

std::unique_ptr<ItemStack> PlayerInventory::getChestplate() const
{
    try {
        return get_stack(impl->inventory->getChestplate());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setChestplate(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setChestplate(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearChestplate() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setChestplate);
}

std::unique_ptr<ItemStack> PlayerInventory::getLeggings() const
{
    try {
        return get_stack(impl->inventory->getLeggings());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setLeggings(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setLeggings(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearLeggings() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setLeggings);
}

std::unique_ptr<ItemStack> PlayerInventory::getBoots() const
{
    try {
        return get_stack(impl->inventory->getBoots());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setBoots(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setBoots(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearBoots() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setBoots);
}

std::unique_ptr<ItemStack> PlayerInventory::getItemInMainHand() const
{
    try {
        return get_stack(impl->inventory->getItemInMainHand());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setItemInMainHand(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setItemInMainHand(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearItemInMainHand() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setItemInMainHand);
}

std::unique_ptr<ItemStack> PlayerInventory::getItemInOffHand() const
{
    try {
        return get_stack(impl->inventory->getItemInOffHand());
    }
    catch (...) {
        return std::unique_ptr<ItemStack>();
    }
}

void PlayerInventory::setItemInOffHand(const ItemStack &item) const
{
    try {
        if (item.native() != nullptr) {
            impl->inventory->setItemInOffHand(std::optional<endstone::ItemStack>{*item.native()});
        }
    }
    catch (...) {
    }
}

void PlayerInventory::clearItemInOffHand() const
{
    clear_stack(*impl->inventory, &endstone::PlayerInventory::setItemInOffHand);
}

} // namespace aegilex::native::inventory

File: rust/crates/aegilex-runtime/src/host/endstone/inventory/player_inventory.rs
//! Core ABI implementation for `native/bindings/endstone/inventory/player_inventory.h`.

use crate::host::endstone::inventory::resources::*;
use crate::host::endstone::support::*;

impl crate::core_host::imports::HostPlayerInventory for PluginStoreState {
    fn player_inventory_get_inventory(
        &mut self,
        self_: u32,
    ) -> Result<Result<u32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-inventory")?;
            let handle =
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?;
            let inventory = self
                .handles
                .player_inventory(self.invocation_id, handle)
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?
                .asInventory();
            self.insert_inventory_resource(inventory)
                .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_get_helmet(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-helmet")?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getHelmet(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_helmet(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.set-helmet")?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setHelmet(item),
                |inventory| inventory.clearHelmet(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_chestplate(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-chestplate")?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getChestplate(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_chestplate(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.set-chestplate")?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setChestplate(item),
                |inventory| inventory.clearChestplate(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_leggings(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-leggings")?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getLeggings(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_leggings(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.set-leggings")?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setLeggings(item),
                |inventory| inventory.clearLeggings(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_boots(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-boots")?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getBoots(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_boots(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.set-boots")?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setBoots(item),
                |inventory| inventory.clearBoots(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_item_in_main_hand(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "player-inventory.player-inventory.get-item-in-main-hand",
            )?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getItemInMainHand(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_item_in_main_hand(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "player-inventory.player-inventory.set-item-in-main-hand",
            )?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setItemInMainHand(item),
                |inventory| inventory.clearItemInMainHand(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_item_in_off_hand(
        &mut self,
        self_: u32,
    ) -> Result<Result<Option<u32>, TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "player-inventory.player-inventory.get-item-in-off-hand",
            )?;
            player_get_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                |inventory| inventory.getItemInOffHand(),
            )
            .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_item_in_off_hand(
        &mut self,
        self_: u32,
        item: Option<u32>,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(
                self,
                "player-inventory.player-inventory.set-item-in-off-hand",
            )?;
            player_set_slot(
                self,
                player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                item,
                |inventory, item| inventory.setItemInOffHand(item),
                |inventory| inventory.clearItemInOffHand(),
            )
            .map_err(map_core_host_error)?;
            Ok(())
        })())
    }

    fn player_inventory_get_held_item_slot(
        &mut self,
        self_: u32,
    ) -> Result<Result<i32, TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.get-held-item-slot")?;
            self.handles
                .player_inventory(
                    self.invocation_id,
                    player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.getHeldItemSlot())
                .ok_or_else(not_found)
                .map_err(map_core_host_error)
        })())
    }

    fn player_inventory_set_held_item_slot(
        &mut self,
        self_: u32,
        slot: i32,
    ) -> Result<Result<(), TypesHostError>, String> {
        Ok((|| {
            check_capability(self, "player-inventory.player-inventory.set-held-item-slot")?;
            self.handles
                .player_inventory(
                    self.invocation_id,
                    player_inventory_resource_handle(self, self_).map_err(map_core_host_error)?,
                )
                .map(|inventory| inventory.setHeldItemSlot(slot))
                .ok_or_else(not_found)
                .map_err(map_core_host_error)?;
            Ok(())
        })())
    }
}
